RCT unpacker plus DQM 74X

nsmith- opened the pull request:
Here is a combo of the RCT unpacker + standard sequence change + DQM updates for L1TRCT and L1TdeRCT.
This PR contains the exact same unpacker as #10220, rebased to 74X.

An identical PR for 75X is #10631
A PR for the subset for 76X that hasn't been merged already
(namely standard sequence + DQM updates)
is #10658

@isobelojalvo, @jingyucms, @mcepeda, you all might be interested in this.

isobelojalvo committed:
committing RCT Raw To Digi

isobelojalvo committed:
committing RCT Raw To Digi

isobelojalvo committed:
Adding in some changes for cmsbuild as well as better error checks

isobelojalvo committed:
Making central BX 2

isobelojalvo committed:
Removing TList from previous glitch

isobelojalvo committed:
Adding in recommended change

isobelojalvo committed:
Adding in recommended change

isobelojalvo committed:
Adding in recommended change

isobelojalvo committed:
Trying again with Tlist

isobelojalvo committed:
Trying again with Tlist

isobelojalvo committed:
Adding in missing build files, thanks mike :)

mcepeda committed:
Add new code for rct unpacker, V0

mcepeda committed:
Changed source

mcepeda committed:
Now RCT&Layer2 unpacked plots

mcepeda committed:
Added central plots

isobelojalvo committed:
adding in a few more checks on data bounds and centering BX at 0

isobelojalvo committed:
Bug fix for HFQbits and deleting unused code

isobelojalvo committed:
Centering BX at 0

Commit:
Merge branch 'updatedDQMForFED1350V2' into RCTDQMupdates

Commit:
Two folder solution to RCT DQM

Commit:
Change folders to align with standalone dqm for RCT

Commit:
Updates to RawToDigi standard sequence for RctToDigi unpacker

cmsbuild commented:
A new Pull Request was created by @nsmith- (Nicholas Smith) for CMSSW_7_4_X.

RCT unpacker plus DQM 74X

It involves the following packages:

Configuration/StandardSequences
DQM/L1TMonitor
EventFilter/RctRawToDigi


@danduggan, @cmsbuild, @franzoni, @deguio, @mulhearn, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @makortel, @GiacomoSguazzoni, @rovere, @VinInn, @Martin-Grunewald, @cerati, @dgulhan this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

deguio commented:
please test
the equivalent of this PR for the DQM part will have to be tested in 76 first. you might want to decouple DQM and unpacker if you need the latter to go in quickly.

ciao,
F.

cmsbuild commented:
The tests are being triggered in jenkins.

Commit:
Remove rctDigis, too ambiguous

+ Delete old, very outdated files

cmsbuild commented:
+1
Tested at: 82cf1829abcfb487b4c1dd73f3a39f61f447184c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10381/6644/summary.html

cmsbuild commented:
Pull request #10381 was updated. @danduggan, @franzoni, @mulhearn, @deguio, @davidlange6 can you please check and sign again.

nsmith- commented on Configuration/StandardSequences/python/RawToDigi_cff.py:
```diff
@@ -55,6 +58,7 @@
 
 RawToDigi = cms.Sequence(csctfDigis
                          +dttfDigis
+                         +l1RctHwDigis
```
It is unclear to me whether this belongs in the sequence here.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10381/6644/summary.html

Commit:
Change name of unpacker to rctHwDigis

mulhearn commented:
@nsmith- the name should be "rctDigis" not "rctHwDigis".  Note that in standard sequences, we always have things like:  
https://github.com/cms-sw/cmssw/blob/CMSSW_7_6_X/L1Trigger/Configuration/python/SimL1Emulator_cff.py#L26
not sure why we do it this way (predates me) but in short you don't have to worry that your unpacked digis will collide with sim digis.  I think this is a case where being different will just confuse matters...

cmsbuild commented:
Pull request #10381 was updated. @danduggan, @cmsbuild, @franzoni, @deguio, @mulhearn, @davidlange6 can you please check and sign again.

isobelojalvo commented:
@richard-cms 
"If 'rctDigis' is already in the 'process' object you might get an error,
but that means that some other version of the unpacker has incorrectly been
included in some sequence and that needs a fix."

OK so then you say @nsmith-   is right, this could cause a problem. Unfortunately, relying on others to configure their CMS SW correctly is not a very safe way to go (as we have seen over the past few years at CMS this very often causes errors on a wide scale). 

@mulhearn , we'll defer to you as the convener.

mulhearn commented:
If we decide this is a lousy convention (currently used everywhere in L1T configs, since before my time) I would rather change all the names in one PR (consistently throughout L1T) instead of just making the change here to this one single module...  so for this PR, definitely stick with the convention!

jingyucms commented:
Hi Nick:
can you put your change regarding 
DQM/L1TMonitor/test/l1t_dqm_sourceclient-file_cfg.py and
DQM/L1TMonitor/test/l1temulator_dqm_sourceclient-file_cfg.py

also into 
DQM/Integration/python/test/l1t_dqm_sourceclient-file_cfg.py and
DQM/Integration/python/test/l1temulator_dqm_sourceclient-file_cfg.py

Thanks and Cheers
Jingyu

isobelojalvo commented:
@jingyucms  I suppose yes, he can. What is the plan for releasing this at P5?

Commit:
Adding requested changes from @jingyucms

nsmith- commented:
Hi @jingyucms 
I have included the changes, although it appears to me that the blocks of code in both files that switch the source to `rawDataCollector` are taken care of in the standard sequence config here:
https://github.com/cms-l1t-offline/cmssw/blob/RCTunpackerPlusDQM74X/Configuration/StandardSequences/python/RawToDigi_cff.py#L95

Either way, better safe than sorry!

cmsbuild commented:
Pull request #10381 was updated. @danduggan, @cmsbuild, @franzoni, @deguio, @mulhearn, @davidlange6 can you please check and sign again.

Commit:
Introduce L1T PU monitoring module to online DQM

Commit:
Tweak positioning

Commit:
Change name of RCT unpacker to rctDigis in Standard Sequence

At the request of L1 offline software folks
For the record, I think this could possibly cause issues in
legacy code, due to the emulator previously being called 'rctDigis'
rather than the preferred 'simRctDigis'

cmsbuild commented:
Pull request #10381 was updated. @danduggan, @cmsbuild, @franzoni, @deguio, @mulhearn, @davidlange6 can you please check and sign again.

deguio commented:
hello @nsmith- 
was a separate PR to 76 for the DQM part submitted already?
thanks,
F.

Commit:
Remove cout

nsmith- commented:
@deguio It was forthcoming until now.
I have updated my top comment, but for reference: #10658

nsmith- commented:
@mulhearn, I have made the change to the producer label to `rctDigis`, can you please re-review this PR?

cmsbuild commented:
Pull request #10381 was updated. @danduggan, @cmsbuild, @franzoni, @deguio, @mulhearn, @davidlange6 can you please check and sign again.

mulhearn commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

mulhearn commented:
+1

cmsbuild commented:
+1
Tested at: 425ab427779652bc46bee9bb3e3b425362d611bd
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10381/7110/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10381/7110/summary.html

Commit:
Change default source of L1 RCT monitoring to caloStage1Digis

Commit:
Revert b603bf081b695ddd46f7b1c53f3046f6f4df6062

Due to merge conflict with 9a95bb204fd57e9bfcd0df681cf0344a0bafdf77

deguio commented:
-1
apart from the cout to be removed, the PR looks ok for what concerns DQM and can go in.
rejecting for now, but feel free to update this branch and the PR will be reconsidered.

keep in mind that this one will need to be merged and succeed a relval cycle in 76 before it can get merged in 74 and 75. 

ciao,
F.

Commit:
Remove cout

cmsbuild commented:
Pull request #10381 was updated. @danduggan, @cmsbuild, @franzoni, @deguio, @mulhearn, @davidlange6 can you please check and sign again.

deguio commented:
please test

deguio commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: efcf614a09bdd7ae6a8f42781b79b05937e4428b
 When I ran the RelVals I found an error in the following worklfows: 
 4.53 step3
<pre>runTheMatrix-results/4.53_RunPhoton2012B+RunPhoton2012B+HLTD+RECODreHLT+HARVESTDreHLT/step3_RunPhoton2012B+RunPhoton2012B+HLTD+RECODreHLT+HARVESTDreHLT.log

----- Begin Fatal Exception 15-Aug-2015 09:40:44 CEST-----------------------

An exception of category 'ProductNotFound' occurred while

   [0] Processing run: 194533 lumi: 329 event: 461582793

   [1] Running path 'dqmoffline_step'

   [2] Calling event method for module L1TdeRCT/'l1TdeRCT'

Exception Message:

Principal::getByToken: Found zero products matching all criteria

Looking for type: std::vector<L1CaloEmCand>

Looking for module label: caloStage1Digis

Looking for productInstanceName: 



   Additional Info:

      [a] If you wish to continue processing events after a ProductNotFound exception,

add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.



----- End Fatal Exception -------------------------------------------------

</pre>
140.53 step2
<pre>runTheMatrix-results/140.53_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI/step2_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI.log

----- Begin Fatal Exception 15-Aug-2015 09:43:58 CEST-----------------------

An exception of category 'ProductNotFound' occurred while

   [0] Processing run: 182124 lumi: 40 event: 1893722

   [1] Running path 'dqmoffline_step'

   [2] Calling event method for module L1TdeRCT/'l1TdeRCT'

Exception Message:

Principal::getByToken: Found zero products matching all criteria

Looking for type: std::vector<L1CaloEmCand>

Looking for module label: caloStage1Digis

Looking for productInstanceName: 



   Additional Info:

      [a] If you wish to continue processing events after a ProductNotFound exception,

add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.



----- End Fatal Exception -------------------------------------------------

</pre>
1000.0 step2
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step2_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

----- Begin Fatal Exception 15-Aug-2015 09:56:47 CEST-----------------------

An exception of category 'ProductNotFound' occurred while

   [0] Processing run: 165121 lumi: 62 event: 23609118

   [1] Running path 'dqmoffline_step'

   [2] Calling event method for module L1TdeRCT/'l1TdeRCT'

Exception Message:

Principal::getByToken: Found zero products matching all criteria

Looking for type: std::vector<L1CaloEmCand>

Looking for module label: caloStage1Digis

Looking for productInstanceName: 



   Additional Info:

      [a] If you wish to continue processing events after a ProductNotFound exception,

add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.



----- End Fatal Exception -------------------------------------------------

</pre>
1001.0 step2
<pre>runTheMatrix-results/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3/step2_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3.log

----- Begin Fatal Exception 15-Aug-2015 09:59:30 CEST-----------------------

An exception of category 'ProductNotFound' occurred while

   [0] Processing run: 165121 lumi: 62 event: 23609118

   [1] Running path 'dqmoffline_step'

   [2] Calling event method for module L1TdeRCT/'l1TdeRCT'

Exception Message:

Principal::getByToken: Found zero products matching all criteria

Looking for type: std::vector<L1CaloEmCand>

Looking for module label: caloStage1Digis

Looking for productInstanceName: 



   Additional Info:

      [a] If you wish to continue processing events after a ProductNotFound exception,

add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.



----- End Fatal Exception -------------------------------------------------

</pre>
1003.0 step2
<pre>runTheMatrix-results/1003.0_RunMinBias2012A+RunMinBias2012A+RECODDQM+HARVESTDDQM/step2_RunMinBias2012A+RunMinBias2012A+RECODDQM+HARVESTDDQM.log

----- Begin Fatal Exception 15-Aug-2015 10:03:02 CEST-----------------------

An exception of category 'ProductNotFound' occurred while

   [0] Processing run: 191226 lumi: 398 event: 556864314

   [1] Running path 'dqmoffline_step'

   [2] Calling event method for module L1TdeRCT/'l1TdeRCT'

Exception Message:

Principal::getByToken: Found zero products matching all criteria

Looking for type: std::vector<L1CaloEmCand>

Looking for module label: caloStage1Digis

Looking for productInstanceName: 



   Additional Info:

      [a] If you wish to continue processing events after a ProductNotFound exception,

add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.



----- End Fatal Exception -------------------------------------------------

</pre>

 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10381/7354/summary.html

danduggan commented:
-1
still needs another commit before ready to go in, same as #10631 and #10658.

mcepeda commented:
Dear Dan, all,
Can we solve this issue? We do not know what  commit is missing.
 "caloStage1Digis"  is in our understanding the unpacked digis from stage1. If we need to use another collection please let us know. 
We understood that the stage1 dqm had taken care of producing these digis already - and they must have had this exact problem with the RelVal, since in 2012 there was no Stage1 FED. 
We can try to work around the issue (even if we do not understand why this is a problem?), but we would like to understand what did the Stage1 DQM do to prevent the RelVal failing.
Any help is very welcom, this is actually very urgent.  This PR should be in the DQM online when we start taking data again.
Regards,
Maria

Commit:
Moving all rctDigis inside customization routines

DQM uses separate paths to differentiate stage1, so made
appropriate additions to stage1 paths

danduggan commented:
Hi @mcepeda, the failures mean that there's something not protected in the code when running the dqm offline step, so even if it's expected to work for future runs, it would be good that it doesn't crash when these digis aren't present. There's an ongoing discussion on this in  #10658. In the meantime, if this is needed for online DQM, we can still add it on top of the release in the online DQM cluster while we converge on this. @jingyucms knows this procedure well :)

mcepeda commented:
Hi @danduggan. We do need this for online DQM, urgently. @jingyucms , can you please help us with that?
Thanks!

danduggan commented:
Hi @mcepeda, thanks for the confirmation. I'm starting the testing for online now (taking the rest of this off git).

cmsbuild commented:
Pull request #10381 was updated. @danduggan, @cmsbuild, @franzoni, @deguio, @mulhearn, @davidlange6 can you please check and sign again.

jingyucms commented:
@danduggan Let me testing on l1t local first.

jingyucms commented:
@danduggan Can you trigger git test again?

danduggan commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 4fe491879056ba07ab53bd4e9faa3669d3967bdd
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10381/7477/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10381/7477/summary.html

mulhearn commented:
+1

nsmith- commented:
Superceded by #10880
